### PR TITLE
DM-13213: Fix linking of dynamic library

### PR DESCRIPTION
**Pull request by SimonKrughoff**
This just removes the link generated by the build and replaces it with a link to the relative path (the file in the same directory).  This will make the package more transportable.

### CODE REVIEW

**Comment by danielsf** [comment]
FYI, building the package `sims_GalSimInterface` should test this.  That is the sims package that calls the GalSim API.  I will start a build when I get home tonight (my mac is not the fast at building the stack).

**Comment by danielsf** [comment]
(thanks for taking this on)

File: ups/eupspkg.cfg.sh [hunk 1]
```diff
@@ -67,4 +67,9 @@ install()
 	default_install
 
 	cp -r include "$PREFIX/"
+	if [[ $OSTYPE == darwin* ]]; then
+		cd lib
+		rm libgalsim.dylib
+		ln -s libgalsim.*.dylib libgalsim.dylib
```
Comment (timj) [inline ups/eupspkg.cfg.sh]: Maybe add `cd ..` after the `ln` so that we don't end up in a different directory if people add things later on.

Comment (SimonKrughoff) [reply ups/eupspkg.cfg.sh]: Good point.

**Comment by SimonKrughoff** [comment]
@danielsf I tried a build on linux and I didn't see the problem, so didn't put a fix in for linux.  I'll trigger a build now.

Edit: Thinking about this more, it doesn't really help to trigger a build on Jenkins because the absolute paths are correct in that environment.  I don't know how to get a binary distribution of this ticket branch.

**Comment by danielsf** [comment]
Yeah... I was mostly suggesting the build to make sure nothing broke.  Actually testing if this fixes the problem will require special intervention from someone who can control the binary build process by hand.

**Comment by timj** [comment]
The other approach is to say "this looks okay in theory and doesn't break the build" and then see on Monday if binaries work... I'd add that if this form of the soft link works for a local build then it should also work for binaries so I'd be fairly confident. Do we know that cmodel works with this in place? I don't think we'll get much traction with upstream because the C++ library is a private interface from their perspective.

File: ups/eupspkg.cfg.sh [hunk 2]
```diff
@@ -67,4 +67,14 @@ install()
 	default_install
 
 	cp -r include "$PREFIX/"
+	if [[ $OSTYPE == darwin* ]]; then
+		curdir=`pwd`
```
Comment (natelust) [inline ups/eupspkg.cfg.sh]: You should use $(pwd) in lieu of backticks

File: ups/eupspkg.cfg.sh [hunk 3]
```diff
@@ -67,4 +67,14 @@ install()
...
+		cd $PREFIX/lib
+		galsim_name=`ls libgalsim.*.dylib`
```
Comment (natelust) [inline ups/eupspkg.cfg.sh]: This is potentially fragile if in the future there happens to be more than one version of the library at once. galsim_name would then be nonsense

File: ups/eupspkg.cfg.sh [hunk 4]
```diff
@@ -67,4 +67,14 @@ install()
...
+		rm libgalsim.dylib
+		ln -s $galsim_name libgalsim.dylib
+		install_name_tool -id $galsim_name $galsim_name
+		install_name_tool -change $PREFIX/lib/$galsim_name @loader_path/../../$galsim_name python/galsim/_galsim.so
```
Comment (natelust) [inline ups/eupspkg.cfg.sh]: Why is this necessary? it should be looking for the symlink you just created. This seems to be making the binary point right to the actual library and not the symlink? Maybe I don't understand the install_name_tool well enough. Can you clarify why you need both this and creating a symlink?

**Review by natelust — CHANGES_REQUESTED** [review]

File: ups/eupspkg.cfg.sh (hunk 3 above)
Comment (SimonKrughoff) [reply ups/eupspkg.cfg.sh]: How would you do this otherwise?  I don't want to have to change this code if the galsim version changes.

File: ups/eupspkg.cfg.sh (hunk 4 above)
Comment (SimonKrughoff) [reply ups/eupspkg.cfg.sh]: I thought the symlink was a convenience for packages outside of galsim.  As far as I can tell, `_galsim.so` always pointed to the versioned library not the symlink.  We needed to specify where the library is relative to `_galsim.so` because it wasn't being picked up in the `@rpath`.  I have to admit that I don't fully understand the Apple linking model, so it's certainly possible that there are other ways to do it, but in my testing this is the first incantation that worked for me.

File: ups/eupspkg.cfg.sh (hunk 2 above)
Comment (SimonKrughoff) [reply ups/eupspkg.cfg.sh]: Good enough.

File: ups/eupspkg.cfg.sh (hunk 3 above)
Comment (natelust) [reply ups/eupspkg.cfg.sh]: `galsim_name_array=( ( ls libgalsim.*.dylib ) )
galsim_name=${galsim_name_array[${#galsim_name_array[@]}-1]}`

This will list the directory, save it as an array, and then select the last element of the array (the length minus 1) This presumes the structure of how the libs are named will not change such that ls will always put then in ascending order, which is probably a reasonable assumption.

This may be something that does not need worried about, I am just trying to guard against an unexpected failure where someone (maybe the build system) does something weird in the future

File: ups/eupspkg.cfg.sh (hunk 4 above)
Comment (natelust) [reply ups/eupspkg.cfg.sh]: I don't fully understand apples linking model either, but I am happy if you are satisfied it is needed and it works.

File: ups/eupspkg.cfg.sh (hunk 3 above)
Comment (timj) [reply ups/eupspkg.cfg.sh]: I'd be inclined to leave it as the simple version since we are in control of when galsim is updated in this package.

Comment (SimonKrughoff) [reply ups/eupspkg.cfg.sh]: I ended up grabbing the version from the python version file.  I think that is the single source of truth for this package.

**Review by natelust — APPROVED** [review]
